File: PerfTweaksBasic/HttpTweak.vb
Imports System.IO
Imports System.Net
Imports System.Security.Cryptography
Imports System.Text
Imports System.Web

Public NotInheritable Class HttpTweak

    Private Const IfNoneMatch As String = "If-None-Match"
    Private Const Etag As String = "Etag"
    Private Const IfModifiedSince As String = "If-Modified-Since"
    Private Shared _dateTime As DateTime

    Public Shared Sub PublicCache(context As HttpContext)
        _dateTime = File.GetLastWriteTime(context.Request.PhysicalPath)
        SetConditionalGetHeaders(context, context.Request.Url.AbsoluteUri)
    End Sub

    Public Shared Sub PrivateCache(context As HttpContext, uri As String)
        SetConditionalGetHeaders(context, uri)
    End Sub

    Private Shared Sub SetConditionalGetHeaders(httpContext As HttpContext, url As String)
        SetConditionalGetHeaders(Hash(url), httpContext)
        SetConditionalGetHeaders(_dateTime, httpContext)
    End Sub

    Private Shared Sub SetConditionalGetHeaders(etag As String, context As HttpContext)
        Dim ifNoneMatchHeader As String = context.Request.Headers(IfNoneMatch)
        etag = """{etag}"""

        If ifNoneMatchHeader IsNot Nothing AndAlso ifNoneMatchHeader.Contains(",") Then
            ifNoneMatchHeader = ifNoneMatchHeader.Substring(0, ifNoneMatchHeader.IndexOf(",", StringComparison.Ordinal))
        End If

        context.Response.AppendHeader(HttpTweak.Etag, etag)
        context.Response.Cache.VaryByHeaders(IfNoneMatch) = True

        If etag = ifNoneMatchHeader Then
            context.Response.ClearContent()
            context.Response.StatusCode = CInt(HttpStatusCode.NotModified)
            context.Response.SuppressContent = True
        End If
    End Sub

    Private Shared Sub SetConditionalGetHeaders(lastModified As DateTime, context As HttpContext)
        Dim response As HttpResponse = context.Response
        Dim request As HttpRequest = context.Request
        lastModified = New DateTime(lastModified.Year, lastModified.Month, lastModified.Day, lastModified.Hour, lastModified.Minute, lastModified.Second)

        Dim incomingDate As String = request.Headers(IfModifiedSince)

        response.Cache.SetLastModified(lastModified)

        Dim testDate As DateTime

        If DateTime.TryParse(incomingDate, testDate) AndAlso testDate = lastModified Then
            response.ClearContent()
            response.StatusCode = CInt(HttpStatusCode.NotModified)
            response.SuppressContent = True
        End If
    End Sub

    Private Shared Function Hash(value As String) As String
        Dim md5 As MD5 = New MD5Cng()

        Dim hashValue As Byte() = md5.ComputeHash(Encoding.UTF8.GetBytes(value))

        Dim stringBuilder = New StringBuilder()

        For index = 0 To hashValue.Length - 2
            stringBuilder.Append(hashValue(index).ToString("x2"))
        Next

        Return stringBuilder.ToString()
    End Function
End Class

File: PerfTweaksLite/HttpTweak.cs
using System;
using System.IO;
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Web;

namespace PerfTweaksLite
{
    public static class HttpTweak
    {
        private const string IfNoneMatch = "If-None-Match";
        private const string Etag = "Etag";
        private const string IfModifiedSince = "If-Modified-Since";
        private static DateTime _dateTime;

        public static void PublicCacheExtension(this HttpContext context)
        {
            _dateTime = File.GetLastWriteTime(context.Request.PhysicalPath);
            SetConditionalGetHeaders(context, context.Request.Url.AbsoluteUri);
        }

        public static void PrivateCacheExtension(this HttpContext context, string uri)
        {
            SetConditionalGetHeaders(context, uri);
        }

        public static void PublicCache(HttpContext context)
        {
            _dateTime = File.GetLastWriteTime(context.Request.PhysicalPath);
            SetConditionalGetHeaders(context, context.Request.Url.AbsoluteUri);
        }

        public static void PrivateCache(HttpContext context, string uri)
        {
            SetConditionalGetHeaders(context, uri);
        }

        private static void SetConditionalGetHeaders(HttpContext httpContext, string url)
        {
            SetConditionalGetHeaders(Hash(url), httpContext);
            SetConditionalGetHeaders(_dateTime, httpContext);
        }

        private static void SetConditionalGetHeaders(string etag, HttpContext context)
        {
            string ifNoneMatch = context.Request.Headers[IfNoneMatch];
            etag = $"\"{etag}\"";

            if (ifNoneMatch != null && ifNoneMatch.Contains(","))
            {
                ifNoneMatch = ifNoneMatch.Substring(0, ifNoneMatch.IndexOf(",", StringComparison.Ordinal));
            }

            context.Response.AppendHeader(Etag, etag);
            context.Response.Cache.VaryByHeaders[IfNoneMatch] = true;

            if (etag == ifNoneMatch)
            {
                context.Response.ClearContent();
                context.Response.StatusCode = (int)HttpStatusCode.NotModified;
                context.Response.SuppressContent = true;
            }
        }

        private static void SetConditionalGetHeaders(DateTime lastModified, HttpContext context)
        {
            HttpResponse response = context.Response;
            HttpRequest request = context.Request;
            lastModified = new DateTime(lastModified.Year, lastModified.Month, lastModified.Day, lastModified.Hour, lastModified.Minute, lastModified.Second);

            string incomingDate = request.Headers[IfModifiedSince];

            response.Cache.SetLastModified(lastModified);

            DateTime testDate;

            if (DateTime.TryParse(incomingDate, out testDate) && testDate == lastModified)
            {
                response.ClearContent();
                response.StatusCode = (int)HttpStatusCode.NotModified;
                response.SuppressContent = true;
            }
        }

        private static string Hash(string value)
        {
            MD5 md5 = new MD5Cng();

            byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(value));

            var stringBuilder = new StringBuilder();

            for (int index = 0; index < hash.Length - 1; ++index)
                stringBuilder.Append(hash[index].ToString("x2"));

            return stringBuilder.ToString();
        }
    }
}
